File: gcc/hls_node/fortran/hls_single.f90
!~         program main
		subroutine mpc_user_main
        
        include 'mpif.h'
        
!      	Variables globales
        integer compteur_node
        save compteur_node

        integer ierror
        integer rank

        !$hls node(compteur_node)
        
        call mpi_init(ierror)
        call mpi_comm_size(MPI_COMM_WORLD, sizes, ierror)
        call mpi_comm_rank(MPI_COMM_WORLD, rank, ierror)

        !$hls barrier(compteur_node)

        !$hls single(compteur_node)
            compteur_node = compteur_node + 1 
        !$hls end single
        
        !$hls barrier(compteur_node)
        
        if(rank .eq. 0) then
!          node
            if(compteur_node .ne. 1) then
                write(6,*) 'HLS NODE SINGLE FAILURE'
                call abort() 
            else
                write(6,*) 'HLS NODE SINGLE SUCCESS'
            endif
        endif
		call mpi_finalize(ierror)
        end

File: compilers/hls/c++/hls_single.cpp
#include <mpi.h>
#include <cstdlib>
#include <iostream>

using namespace std;

int compteur_node = 0;

#pragma hls node(compteur_node)

int main( int argc, char **argv )
{
	int rank, size;
	MPI_Status status;
	MPI_Request request;
	MPI_Init( &argc, &argv );
	MPI_Comm_rank( MPI_COMM_WORLD, &rank );
	MPI_Comm_size( MPI_COMM_WORLD, &size );

	#pragma hls barrier(compteur_node)
	
	#pragma hls single(compteur_node)
	{
		compteur_node++;
	}

	#pragma hls barrier(compteur_node)
	
	if(rank == 0)
	{
		//node
		if(compteur_node != 1)
		{
			cout << "HLS NODE SINGLE FAILURE\n" << endl;
			abort();
		}
		else
		{
			cout << "HLS NODE SINGLE SUCCESS\n" << endl;
		}
	}
	
	MPI_Finalize();
	return EXIT_SUCCESS;
}
